src/main/java/org/rumbledb/expressions/module/Prolog.java

```diff
@@ -27,42 +27,43 @@
 
 import java.util.ArrayList;
 import java.util.List;
+import java.util.stream.Collectors;
 
 public class Prolog extends Node {
 
-    private final List<VariableDeclaration> variableDeclarations;
-    private final List<FunctionDeclaration> functionDeclarations;
+    private List<Node> declarations;
```

Comment:
I do not exactly see the benefit of this refactoring at the moment. If we are planning to add more declaration types it seems appealing to not have separate lists for each type. However, it seems that the declaration type information is available in the parameters of the constructor. This is then lost, only to be recalculated by the getter methods. This feels redundant from a computational perspective. Is there something that I'm missing?

Reply:
This is because the order in this list is important, and it may alternate between variables and functions.

Example:

1. declare variable $x := 1;
2. declare function fn($i) { $i + $x };
3. declare variable $y := fn(2);

The variable dependency visitor will construct a DAG with all declarations, and then re-orders this list in a way compatible with the dependencies, so that the dynamic context are built in the correct order.

Reply:
> This is because the order in this list is important, and it may alternate between variables and functions.

This was exactly what I needed. Thanks for the detailed explanation!